// nv-file-const/index.js
const TNAMES = {
    slink:"slink",
    valid_slink:"valid_slink",
    invalid_slink:"invalid_slink",
    file:"file",
    dir:"dir",
    blk_dev:"blk_dev",
    char_dev:"char_dev",
    fifo:"fifo",
    sock:"sock",
    noexist:"noexist"
}


const CONST_TNAME_TO_NUM  = {
  S_IFMT:  0b1111000000000000,
  S_IFREG: 0b1000000000000000,      //8
  S_IFDIR: 0b0100000000000000,      //4
  S_IFCHR: 0b0010000000000000,      //2
  S_IFBLK: 0b0110000000000000,      //6
  S_Ififo: 0b0001000000000000,      //1
  S_IFLNK: 0b1010000000000000,      //10
  S_IFSOCK:0b1100000000000000,      //12
}

const NUM_TO_CONST_TNAME = {
    "4096": "S_Ififo",
    "8192": "S_IFCHR",
    "16384": "S_IFDIR",
    "24576": "S_IFBLK",
    "32768": "S_IFREG",
    "40960": "S_IFLNK",
    "49152": "S_IFSOCK",
    "61440": "S_IFMT"
}

const CONST_TNAME_TO_TNAME = {
    S_IFREG: "file",
    S_IFDIR: "dir",
    S_IFCHR: "char_dev",
    S_IFBLK: "blk_dev",
    S_Ififo: "fifo",
    S_IFLNK: "slink",
    S_IFSOCK: "sock"
}

const TNAME_TO_CONST_TNAME = {
    "file": "S_IFREG",
    "dir": "S_IFDIR",
    "char_dev": "S_IFCHR",
    "blk_dev": "S_IFBLK",
    "fifo": "S_Ififo",
    "slink": "S_IFLNK",
    "sock": "S_IFSOCK"
}

const num_to_tname = (n) => CONST_TNAME_TO_TNAME[NUM_TO_CONST_TNAME[n]];

const tname_to_num = (tname) => CONST_TNAME_TO_NUM[TNAME_TO_CONST_TNAME[tname]];


const PNAMES = {
    owner_can_read:'owner_can_read',
    owner_can_write:'owner_can_write',
    owner_can_exec:'owner_can_exec',
    ////
    group_can_read:'group_can_read',
    group_can_write:'group_can_write',
    group_can_exec:'group_can_exec',
    ////
    other_can_read:'other_can_read',
    other_can_write:'other_can_write',
    other_can_exec:'other_can_exec',
}

const CONST_PNAME_TO_PNAME = {
    S_IRUSR:'owner_can_read',
    S_IWUSR:'owner_can_write',
    S_IXUSR:'owner_can_exec',
    ////
    S_IRGRP:'group_can_read',
    S_IWGRP:'group_can_write',
    S_IXGRP:'group_can_exec',
    ////
    S_IROTH:'other_can_read',
    S_IWOTH:'other_can_write',
    S_IXOTH:'other_can_exec',
}

const PNAME_TO_CONST_PNAME = {
    "owner_can_read": "S_IRUSR",
    "owner_can_write": "S_IWUSR",
    "owner_can_exec": "S_IXUSR",
    "group_can_read": "S_IRGRP",
    "group_can_write": "S_IWGRP",
    "group_can_exec": "S_IXGRP",
    "other_can_read": "S_IROTH",
    "other_can_write": "S_IWOTH",
    "other_can_exec": "S_IXOTH"
}

const CONST_PNAME_TO_NUM = {
  S_IRUSR: 0b0100000000,
  S_IWUSR: 0b0010000000,
  S_IXUSR: 0b0001000000,
  
  S_IRGRP: 0b0000100000,
  S_IWGRP: 0b0000010000,
  S_IXGRP: 0b0000001000,
  

  S_IROTH: 0b0000000100,
  S_IWOTH: 0b0000000010,
  S_IXOTH: 0b0000000001
}

const NUM_TO_CONST_PNAME = {
    "1": "S_IXOTH",
    "2": "S_IWOTH",
    "4": "S_IROTH",
    "8": "S_IXGRP",
    "16": "S_IWGRP",
    "32": "S_IRGRP",
    "64": "S_IXUSR",
    "128": "S_IWUSR",
    "256": "S_IRUSR"
}

const num_to_pname = (n) => CONST_PNAME_TO_PNAME[NUM_TO_CONST_PNAME[n]];

const pname_to_num = (pname) => CONST_PNAME_TO_NUM[PNAME_TO_CONST_PNAME[pname]];


module.exports = {
    TNAMES,
    ////
    CONST_TNAME_TO_NUM,
    NUM_TO_CONST_TNAME,
    ////
    CONST_TNAME_TO_TNAME,
    TNAME_TO_CONST_TNAME,
    ////
    num_to_tname,
    tname_to_num,
    ////
    PNAMES,
    CONST_PNAME_TO_PNAME,
    PNAME_TO_CONST_PNAME,
    ////
    CONST_PNAME_TO_NUM,
    NUM_TO_CONST_PNAME,
    ////
    num_to_pname,
    pname_to_num,
}
