// nv-buf-serde/const.js
const {_u,_n,_t,_f} = require("nv-facutil-untf");


const SerializationTag = {
  kVersion: 255,
  kPadding: 0,
  kVerifyObjectCount: 63,
  kTheHole: 45,
  kUndefined: 95,
  kNull: 48,
  kTrue: 84,
  kFalse: 70,
  kInt32: 73,
  kUint32: 85,
  kDouble: 78,
  kBigInt: 90,
  kUtf8String: 83,
  kOneByteString: 34,
  kTwoByteString: 99,
  kObjectReference: 94,
  kBeginJSObject: 111,
  kEndJSObject: 123,
  kBeginSparseJSArray: 97,
  kEndSparseJSArray: 64,
  kBeginDenseJSArray: 65,
  kEndDenseJSArray: 36,
  kDate: 68,
  kTrueObject: 121,
  kFalseObject: 120,
  kNumberObject: 110,
  kBigIntObject: 122,
  kStringObject: 115,
  kRegExp: 82,
  kBeginJSMap: 59,
  kEndJSMap: 58,
  kBeginJSSet: 39,
  kEndJSSet: 44,
  kArrayBuffer: 66,                            //B 0x42
  kResizableArrayBuffer: 126,
  kArrayBufferTransfer: 116,                   // unsupport 
  kArrayBufferView: 86,                        //V 0x56 --------------- did not use , use nodejs delegate  kHostObject + 
  kSharedArrayBuffer: 117,                     // unsupport
  kSharedObject: 112,                          // unsupport
  kWasmModuleTransfer: 119,                    // unsupport
  kHostObject: 92,
  kWasmMemoryTransfer: 109,                   //unsupport 
  kError: 114,
  kLegacyReservedMessagePort: 77,
  kLegacyReservedBlob: 98,
  kLegacyReservedBlobIndex: 105,
  kLegacyReservedFile: 102,
  kLegacyReservedFileIndex: 101,
  kLegacyReservedDOMFileSystem: 100,
  kLegacyReservedFileList: 108,
  kLegacyReservedFileListIndex: 76,
  kLegacyReservedImageData: 35,
  kLegacyReservedImageBitmap: 103,
  kLegacyReservedImageBitmapTransfer: 71,
  kLegacyReservedOffscreenCanvas: 72,
  kLegacyReservedCryptoKey: 75,
  kLegacyReservedRTCCertificate: 107
}

const V8_ArrayBufferViewTag = {
  kInt8Array: 98,
  kUint8Array: 66,
  kUint8ClampedArray: 67,
  kInt16Array: 119,
  kUint16Array: 87,
  kInt32Array: 100,
  kUint32Array: 68,
  kFloat32Array: 102,
  kFloat64Array: 70,
  kBigInt64Array: 113,
  kBigUint64Array: 81,
  kDataView: 63
}
Object.defineProperty(V8_ArrayBufferViewTag,"UNUSED_",{get:function(){return(_t)}})


const kNodeFastBufferSym = Symbol("#kNodeFastBuffer");

const ArrayBufferViewTag = {
  kInt8Array          :  0,
  kUint8Array         : 1,
  kUint8ClampedArray  : 2,
  kInt16Array         : 3,
  kUint16Array    : 4,
  kInt32Array     : 5,
  kUint32Array    : 6,
  kFloat32Array   : 7,
  kFloat64Array   : 8,
  kDataView       : 9,  
  kNodeFastBuffer   : 10,      //for compatible with node-js-fast-buffer, treated-as kUint8Array just add a 
  kBigInt64Array:  11,
  kBigUint64Array: 12,
}

class NodeFastBuffer extends Uint8Array { 
    [kNodeFastBufferSym] = _t;
}

const IsNodeFastBuffer = (o)=>o[kNodeFastBufferSym] === _t;

const  ObjectPrototypeToString = (o) => Object.prototype.toString.call(o);


//<copied FROM nodejs/lib> 
const  arrayBufferViewTypeToIndex  = (abView) => {
  const type = ObjectPrototypeToString(abView);
  if (type === '[object Int8Array]') return 0;
  if (type === '[object Uint8Array]') return 1;
  if (type === '[object Uint8ClampedArray]') return 2;
  if (type === '[object Int16Array]') return 3;
  if (type === '[object Uint16Array]') return 4;
  if (type === '[object Int32Array]') return 5;
  if (type === '[object Uint32Array]') return 6;
  if (type === '[object Float32Array]') return 7;
  if (type === '[object Float64Array]') return 8;
  if (type === '[object DataView]') return 9;
  // Index 10 is FastBuffer.
  if (type === '[object BigInt64Array]') return 11;
  if (type === '[object BigUint64Array]') return 12;
  return -1;
}
//</copied FROM nodejs/lib >

const ArrayBufferViewIndexToCtor = [
    Int8Array,
    Uint8Array,
    Uint8ClampedArray,
    Int16Array,
    Uint16Array,
    Int32Array,
    Uint32Array,
    Float32Array,
    Float64Array,
    DataView,
    NodeFastBuffer,
    BigInt64Array,
    BigUint64Array
];


const ErrorTag  = {
  kEvalErrorPrototype: 69,
  kRangeErrorPrototype: 82,
  kReferenceErrorPrototype: 70,
  kSyntaxErrorPrototype: 83,
  kTypeErrorPrototype: 84,
  kUriErrorPrototype: 85,
  kMessage: 109,
  kCause: 99,
  kStack: 115,
  kEnd: 46
} 

const kCantBeSered           =  "#___kCantBeSered___";
const CreatFakeCantBeSered   = (id,hint)=>({id,kCantBeSered:hint});

const kUnknown               =  "#___kUnknown___";
const CreatUnknown           = ()=>({kUnknown:_u});


const CODEC_NAME = "v8ser";

module.exports = {
   CODEC_NAME, kNodeFastBufferSym,
   IsNodeFastBuffer,
   //// 
   SerializationTag,
   V8_ArrayBufferViewTag,  //unused
   ArrayBufferViewTag, ArrayBufferViewIndexToCtor, arrayBufferViewTypeToIndex, 
   ErrorTag, 
   ////
   kCantBeSered,
   CreatFakeCantBeSered, 
   ////
   kUnknown,
   CreatUnknown,	
}
